File: src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2020 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.extension.commentary;

import com.intellij.codeInsight.actions.MultiCaretCodeInsightActionHandler;
import com.intellij.codeInsight.generation.CommentByBlockCommentHandler;
import com.intellij.codeInsight.generation.CommentByLineCommentHandler;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.application.WriteAction;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiFile;
import com.maddyhome.idea.vim.VimPlugin;
import com.maddyhome.idea.vim.command.CommandState;
import com.maddyhome.idea.vim.command.MappingMode;
import com.maddyhome.idea.vim.command.SelectionType;
import com.maddyhome.idea.vim.common.TextRange;
import com.maddyhome.idea.vim.extension.VimExtension;
import com.maddyhome.idea.vim.extension.VimExtensionHandler;
import com.maddyhome.idea.vim.key.OperatorFunction;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static com.maddyhome.idea.vim.extension.VimExtensionFacade.*;
import static com.maddyhome.idea.vim.helper.StringHelper.parseKeys;

/**
 * @author dhleong
 */
public class CommentaryExtension implements VimExtension {

  @Override
  public @NotNull String getName() {
    return "commentary";
  }

  @Override
  public void init() {
    putExtensionHandlerMapping(MappingMode.N, parseKeys("<Plug>(CommentMotion)"), getOwner(), new CommentMotionHandler(), false);
    putExtensionHandlerMapping(MappingMode.N, parseKeys("<Plug>(CommentLine)"), getOwner(), new CommentLineHandler(), false);
    putExtensionHandlerMapping(MappingMode.XO, parseKeys("<Plug>(CommentMotionV)"), getOwner(), new CommentMotionVHandler(), false);

    putKeyMapping(MappingMode.N, parseKeys("gc"), getOwner(), parseKeys("<Plug>(CommentMotion)"), true);
    putKeyMapping(MappingMode.N, parseKeys("gcc"), getOwner(), parseKeys("<Plug>(CommentLine)"), true);
    putKeyMapping(MappingMode.XO, parseKeys("gc"), getOwner(), parseKeys("<Plug>(CommentMotionV)"), true);
  }

  private static class CommentMotionHandler implements VimExtensionHandler {
    @Override
    public boolean isRepeatable() {
      return true;
    }

    @Override
    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
      setOperatorFunction(new Operator());
      executeNormalWithoutMapping(parseKeys("g@"), editor);
    }
  }

  private static class CommentMotionVHandler implements VimExtensionHandler {
    @Override
    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
      if (!editor.getCaretModel().getPrimaryCaret().hasSelection()) {
        return;
      }

      // always use line-wise comments
      if (!new Operator().apply(editor, context, SelectionType.LINE_WISE)) {
        return;
      }

      WriteAction.run(() -> {
        // Leave visual mode
        executeNormalWithoutMapping(parseKeys("<Esc>"), editor);
        editor.getCaretModel().moveToOffset(editor.getCaretModel().getPrimaryCaret().getSelectionStart());
      });
    }
  }

  private static class Operator implements OperatorFunction {
    @Override
    public boolean apply(@NotNull Editor editor, @NotNull DataContext context, @NotNull SelectionType selectionType) {
      final TextRange range = getCommentRange(editor);
      if (range == null) return false;

      if (CommandState.getInstance(editor).getMode() != CommandState.Mode.VISUAL) {
        editor.getSelectionModel().setSelection(range.getStartOffset(), range.getEndOffset());
      }

      final MultiCaretCodeInsightActionHandler handler =
        selectionType == SelectionType.CHARACTER_WISE
          ? new CommentByBlockCommentHandler()
          : new CommentByLineCommentHandler();

      return WriteAction.compute(() -> {
        try {
          Project proj = editor.getProject();
          if (proj == null) return false;

          PsiFile file = PsiDocumentManager.getInstance(proj).getPsiFile(editor.getDocument());
          if (file == null) return false;

          handler.invoke(editor.getProject(), editor, editor.getCaretModel().getCurrentCaret(), file);
          handler.postInvoke();

          // Jump back to start if in block mode
          if (selectionType == SelectionType.CHARACTER_WISE) {
            executeNormalWithoutMapping(parseKeys("`["), editor);
          }
          return true;
        } finally {
          // remove the selection
          editor.getSelectionModel().removeSelection();
        }
      });
    }

    private @Nullable TextRange getCommentRange(@NotNull Editor editor) {
      final CommandState.Mode mode = CommandState.getInstance(editor).getMode();
      switch (mode) {
        case COMMAND:
          return VimPlugin.getMark().getChangeMarks(editor);
        case VISUAL:
          Caret primaryCaret = editor.getCaretModel().getPrimaryCaret();
          return new TextRange(primaryCaret.getSelectionStart(), primaryCaret.getSelectionEnd());
        default:
          return null;
      }
    }
  }

  private static class CommentLineHandler implements VimExtensionHandler {
    @Override
    public boolean isRepeatable() {
      return true;
    }

    @Override
    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
      final int offset = editor.getCaretModel().getOffset();
      final int line = editor.getDocument().getLineNumber(offset);
      final int lineStart = editor.getDocument().getLineStartOffset(line);
      final int lineEnd = editor.getDocument().getLineEndOffset(line);
      VimPlugin.getMark().setChangeMarks(editor, new TextRange(lineStart, lineEnd));
      new Operator().apply(editor, context, SelectionType.LINE_WISE);
    }
  }
}

File: src/main/java/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.kt
/*
 * Copyright 2003-2026 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.extension.commentary

import com.intellij.openapi.editor.impl.editorId
import com.intellij.vim.api.VimInitApi
import com.maddyhome.idea.vim.KeyHandler
import com.maddyhome.idea.vim.api.ExecutionContext
import com.maddyhome.idea.vim.api.ImmutableVimCaret
import com.maddyhome.idea.vim.api.VimEditor
import com.maddyhome.idea.vim.api.globalOptions
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.command.MappingMode
import com.maddyhome.idea.vim.command.OperatorArguments
import com.maddyhome.idea.vim.command.TextObjectVisualType
import com.maddyhome.idea.vim.common.CommandAliasHandler
import com.maddyhome.idea.vim.common.TextRange
import com.maddyhome.idea.vim.ex.ranges.Range
import com.maddyhome.idea.vim.ex.ranges.toTextRange
import com.maddyhome.idea.vim.extension.ExtensionHandler
import com.maddyhome.idea.vim.extension.VimExtension
import com.maddyhome.idea.vim.extension.VimExtensionFacade
import com.maddyhome.idea.vim.extension.VimExtensionFacade.addCommand
import com.maddyhome.idea.vim.extension.VimExtensionFacade.executeNormalWithoutMapping
import com.maddyhome.idea.vim.extension.VimExtensionFacade.putExtensionHandlerMapping
import com.maddyhome.idea.vim.extension.VimExtensionFacade.putKeyMapping
import com.maddyhome.idea.vim.extension.VimExtensionFacade.putKeyMappingIfMissing
import com.maddyhome.idea.vim.extension.exportOperatorFunction
import com.maddyhome.idea.vim.group.comment.CommentaryRemoteApi
import com.maddyhome.idea.vim.group.rpc
import com.maddyhome.idea.vim.handler.TextObjectActionHandler
import com.maddyhome.idea.vim.key.OperatorFunction
import com.maddyhome.idea.vim.newapi.ij
import com.maddyhome.idea.vim.state.mode.SelectionType

internal class CommentaryExtension : VimExtension {

  object Util {
    fun doCommentary(
      editor: VimEditor,
      context: ExecutionContext,
      range: TextRange,
      selectionType: SelectionType,
      resetCaret: Boolean = true,
    ): Boolean {
      val mode = editor.mode
      val ijEditor = editor.ij
      val document = ijEditor.document

      val editorId = ijEditor.editorId()
      val caretOffset = if (resetCaret) range.startOffset else -1
      if (selectionType === SelectionType.LINE_WISE) {
        val startLine = document.getLineNumber(range.startOffset)
        var endLine = document.getLineNumber(range.endOffset)
        // Adjust endLine if the range ends at the start of a line (don't include that line)
        if (endLine > startLine && document.getLineStartOffset(endLine) == range.endOffset) {
          endLine--
        }
        val finalEndLine = endLine
        rpc(ijEditor.project) {
          CommentaryRemoteApi.getInstance().toggleLineComment(editorId, startLine, finalEndLine, caretOffset)
        }
      } else {
        rpc(ijEditor.project) {
          CommentaryRemoteApi.getInstance()
            .toggleBlockComment(editorId, range.startOffset, range.endOffset, caretOffset)
        }
      }
      return true
    }
  }

  companion object {
    private const val OPERATOR_FUNC = "CommentaryOperatorFunc"
  }

  override fun getName() = "commentary"

  override fun init(initApi: VimInitApi) {
    val plugCommentaryKeys = injector.parser.parseKeys("<Plug>Commentary")
    val plugCommentaryLineKeys = injector.parser.parseKeys("<Plug>CommentaryLine")
    putExtensionHandlerMapping(MappingMode.NX, plugCommentaryKeys, owner, CommentaryOperatorHandler(), false)
    putExtensionHandlerMapping(MappingMode.O, plugCommentaryKeys, owner, CommentaryMappingHandler(), false)
    putKeyMappingIfMissing(MappingMode.N, plugCommentaryLineKeys, owner, injector.parser.parseKeys("gc_"), true)

    putKeyMappingIfMissing(MappingMode.NXO, injector.parser.parseKeys("gc"), owner, plugCommentaryKeys, true)
    putKeyMappingIfMissing(MappingMode.N, injector.parser.parseKeys("gcc"), owner, plugCommentaryLineKeys, true)
    putKeyMappingIfMissing(
      MappingMode.N,
      injector.parser.parseKeys("gcu"),
      owner,
      injector.parser.parseKeys("<Plug>Commentary<Plug>Commentary"),
      true,
    )

    // Previous versions of IdeaVim used different mappings to Vim's Commentary. Make sure everything works if someone
    // is still using the old mapping
    putKeyMapping(MappingMode.N, injector.parser.parseKeys("<Plug>(CommentMotion)"), owner, plugCommentaryKeys, true)
    putKeyMapping(MappingMode.XO, injector.parser.parseKeys("<Plug>(CommentMotionV)"), owner, plugCommentaryKeys, true)
    putKeyMapping(MappingMode.N, injector.parser.parseKeys("<Plug>(CommentLine)"), owner, plugCommentaryLineKeys, true)

    addCommand("Commentary", CommentaryCommandAliasHandler())

    VimExtensionFacade.exportOperatorFunction(OPERATOR_FUNC, CommentaryOperatorFunction())
  }

  private class CommentaryOperatorFunction : OperatorFunction {
    // todo make it multicaret
    override fun apply(editor: VimEditor, context: ExecutionContext, selectionType: SelectionType?): Boolean {
      val range = injector.markService.getChangeMarks(editor.primaryCaret()) ?: return false
      return Util.doCommentary(editor, context, range, selectionType ?: SelectionType.CHARACTER_WISE, true)
    }
  }

  /**
   * Sets up the operator, pending a motion
   *
   * E.g. handles the `gc` in `gc_`, by setting the operator function, then invoking `g@` to receive the `_` motion to
   * invoke the operator. This object is both the mapping handler and the operator function.
   */
  private class CommentaryOperatorHandler : ExtensionHandler {
    override val isRepeatable = true

    override fun execute(editor: VimEditor, context: ExecutionContext, operatorArguments: OperatorArguments) {
      injector.globalOptions().operatorfunc = OPERATOR_FUNC
      executeNormalWithoutMapping(injector.parser.parseKeys("g@"), editor.ij)
    }
  }

  private class CommentaryMappingHandler : ExtensionHandler {
    override val isRepeatable = true

    override fun execute(editor: VimEditor, context: ExecutionContext, operatorArguments: OperatorArguments) {
      val keyState = KeyHandler.getInstance().keyHandlerState
      keyState.commandBuilder.addAction(CommentaryTextObjectMotionHandler)
    }
  }

  /**
   * The text object handler that provides the motion in e.g. `dgc`
   *
   * Delegates to [VimPsiService.getCommentBlockRange][com.maddyhome.idea.vim.api.VimPsiService.getCommentBlockRange]
   * which uses PSI on the backend to detect contiguous comment lines.
   */
  private object CommentaryTextObjectMotionHandler : TextObjectActionHandler() {
    override val visualType: TextObjectVisualType = TextObjectVisualType.LINE_WISE

    override fun getRange(
      editor: VimEditor,
      caret: ImmutableVimCaret,
      context: ExecutionContext,
      count: Int,
      rawCount: Int,
    ): TextRange? {
      return injector.psiService.getCommentBlockRange(editor, caret.getBufferPosition().line)
    }
  }

  /**
   * The handler for the `Commentary` user defined command
   *
   * Used like `:1,3Commentary` or `g/fun/Commentary`
   */
  private class CommentaryCommandAliasHandler : CommandAliasHandler {
    override fun execute(command: String, range: Range, editor: VimEditor, context: ExecutionContext) {
      Util.doCommentary(
        editor,
        context,
        range.getLineRange(editor, editor.primaryCaret()).toTextRange(editor),
        SelectionType.LINE_WISE,
        false
      )
    }
  }
}
